odoo/orm/fields_properties.py

```diff
@@ -29,6 +29,54 @@ def check_property_field_value_name(property_name):
     if not (0 < len(property_name) <= 512) or not regex_alphanumeric.match(property_name):
         raise ValueError(f"Wrong property field value name {property_name!r}.")
 
+class Property(dict):
+    """Allow to be used in `convert_to_record` to make the verification only if we try accessing the property value."""
+
+    def __init__(self, values, field, record):
+        super().__init__(values)
+        self.record = record
+        self.field = field
+
+    def __getitem__(self, property_name):
```

Comment:
Note that `dict` is not an abstract implementation.  The method `dict.get` does not call `dict.__getitem__`; it will therefore leak its "inner" value.  The same holds for methods `dict.values()`, etc.  The class `Property` should subclass `collections.abc.Mapping`, and store the cached value inside a "private" attribute.  What you want is a proxy object to the dict, not the dict itself.

Reply:
the onchange behave weirdly with a `collections.abc.Mapping`, I'll check if I have enough time when we know what approach we take (and when I did it)